piff/gp_interp.py

```diff
@@ -0,0 +1,324 @@
+        optimizer = self.gp.optimizer
+        self.gp.optimizer = None
+        self._fit(data['X'][0], data['Y'][0])
+        self.gp.optimizer = optimizer
+        # Now that gp is setup, we can restore it's initial kernel.
+        self.gp.kernel.theta = self.kernel.theta = data['INIT_THETA'][0]
+
+
+
+from sklearn.gaussian_process.kernels import StationaryKernelMixin, NormalizedKernelMixin, Kernel
+from sklearn.gaussian_process.kernels import Hyperparameter
+
+class EmpiricalKernel(StationaryKernelMixin, NormalizedKernelMixin, Kernel):
+    """ A GaussianProcessRegressor Kernel that just wraps an arbitrary python function.
+
+    :param  fn:  Python callable that accepts two numpy arrays indicating the differences in two
+                 covariates and returns a numpy array indicating the covariances implied by those
+                 differences.
+    """
+    def __init__(self, fn):
+        self.fn = fn
+        assert self.fn(0, 0) == 1
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        if eval_gradient:
+            raise RuntimeError("Cannot evaluate gradient for EmpiricalKernel.")
+
+        X = np.atleast_2d(X)
+        if Y is None:
+            Y = X
+
+        # Only writen for 2D covariance at the moment
+        xshift = np.subtract.outer(X[:,0], Y[:,0])
+        yshift = np.subtract.outer(X[:,1], Y[:,1])
+        return self.fn(xshift, yshift)
+
+
+class AnisotropicRBF(StationaryKernelMixin, NormalizedKernelMixin, Kernel):
+    """ A GaussianProcessRegressor Kernel representing a radial basis function (essentially a
+    squared exponential or Gaussian) but with arbitrary anisotropic covariance.
+
+    While the parameter for this kernel is an inverse covariance function, for the purposes of
+    optimization, it's necessary to reparameterize in such a way to ensure that the covariance
+    matrix is always positive definite.  To this end, we define `theta` (abbreviated `th` below)
+    such that
+
+    invLam = L * L.T
+    L = [[exp(th[0])  0              0           ...    0                 0           ]
+          th[n]       exp(th[1])]    0           ...    0                 0           ]
+          th[n+1]     th[n+2]        exp(th[3])  ...    0                 0           ]
+          ...         ...            ...         ...    ...               ...         ]
+          th[]        th[]           th[]        ...    exp(th[n-2])      0           ]
+          th[]        th[]           th[]        ...    th[n*(n+1)/2-1]   exp(th[n-1])]]
+
+    I.e., the inverse covariance matrix is Cholesky-decomposed, exp(theta[0:n]) lie
+    on the diagonal of the Cholesky matrix, and theta[n:n*(n+1)/2] lie in the lower triangular
+    part of the Cholesky matrix.  This parameterization invertably maps all valid n x n
+    covariance matrices to R^(n*(n+1)/2).  I.e., the range of each theta[i] is -inf ... inf.
+
+    :param  invLam:  Inverse covariance matrix of radial basis function.
```

Comment:
>How is the user supposed to specify this in the config file? We need to have Piff create the invLam function for them based on some measurement of the data.

My thought on this one was that the user would just supply an initial `invLam`, and then `GPInterp.fit` would optimize this (if `optimizer != None`) based on the data.  (Can the config yaml handle a 2x2 numpy array, or list of lists, as a parameter?) There's nothing in this PR currently for generating a good initial guess, though that's certainly a reasonable idea.